src/GitHub.Api/Cache/CachingClasses.cs

```diff
@@ -0,0 +1,18 @@
+using GitHub.Logging;
+using GitHub.Unity;
+using System;
+using System.Collections.Generic;
+using System.Globalization;
+using System.Linq;
+using System.Runtime.Serialization;
+
+namespace GitHub.Unity
+{
+    sealed class RepositoryInfoCacheData : IRepositoryInfoCacheData
```

Comment:
Are you making a single unit out of these values? Is this a pattern we should be repeating in the cache classes?

Reply:
There's two reasons I'm doing this. The first reason is the major one:

When updating fields in the cache, every individual field set will save it to disk and it will raise the cache updated event. Now each cache object has multiple fields, and they're all updated at the same time when the data comes in from git, which means every set will trigger a save and a cache update event. All these events coming from a single cache object will contain partially updated information until the very last field update. The views will get these events and render them multiple times. That's not ideal.

Given that a cache object has multiple pieces of data but represents one atomic event unit, then each update to a cache object must be made atomically with all the information we have for it in one go, and then saved as one piece and one update event raised from it when it's done.

To avoid raising multiple events per cache update and still have the cache save itself and set its updatedDate fields properly, we can't update its fields one by one, we have to update it all in one go (if we set fields individually, then the cache has no way of knowing when the update is done, which means the caller is now responsible for manually telling the cache that the update is complete, which is an extra burden that would be nice not to have).

That means that we have to gather all the information and pass it in a single blob to the cache object, for it to update itself in one go and raise a single event. For that to work in a nice way, we have to have a way of representing the blob of data that's inside the cache. This is what I'm doing here as an experiment: splitting the functionality and the data of the RepositoryInfoCache, so that we can have objects that represent only the data inside it, and we don't have to have separate objects to represent this data (or some other untyped data structure to represent blobs of data).


The other reason is that in order to test things properly I have to have properly working cache objects on the testing side. Since they're all scriptable objects, I can't reference them directly in the standalone tests, which is really really annoying.  I could create some Unity tests, but we can't run those in CI (they run in Unity itself).

So I ended up copy pasting all of the cache objects into IntegrationTests and removing all the Unity specific stuff from them, which is a paaaaaaain 'cause now we have to keep them in sync. We can't use the trick of subclassing them from GitHub.Api because they already subclass ScriptableObject. Sooooo, I'm trying to find ways to pull out as much as possible from the cache objects into GitHub.Api and only leave the bare minimum in the cache objects in GitHub.Unity. Maybe all they do is redirect all their calls into another object that implements all the logic and lives somewhere we can use in tests? Not sure yet, but this is what this is experimenting with.

Reply:
If this works, yeah, I'll change all the other cache classes to use this pattern.